File: 5kyu/valid_parentheses.cpp
#include <stdbool.h>
#include <stdio.h>

bool validParentheses(const char* strin) {
	char	stack[50];
	int		top = 0;

	for (int i = 0; strin[i]; i++) {
		if (strin[i] == '(') {
			stack[top] = strin[i];
			top++;
		}
		if (strin[i] == ')') {
			if (top == 0)
				return false;
			top--;
			if (stack[top] != '(')
				return false;
		}
	}
	if (top != 0)
		return false;
	return true;
}

int		main()
{
	printf("%d", validParentheses("hi())("));
}

File: 5kyu/valid_parentheses.py
def valid_parentheses(string):
    top = 0
    stack = ''
    for char in string:
        if char == '(':
            stack += char
            top += 1
        if char == ')':
            if top == 0:
                return False
            top -= 1
            if not stack[top] == '(':
                return False
    if top == 0:
        return True
    else:
        return False

print(valid_parentheses("hi())("))